time lookup formulas working for pseudoyears

willbutler42 opened the pull request:
If using pseudoyears, the lookup formulas generated in `g3s_time` can be scrambled. For instance, if the year is 40 and the step is 1, `g3s_time_convert` will produce 401, `g3s_time_multiplier` will subsequently produce 1L which removes `cur_step` from the lookup formula. 

The changes make the multiplier (used in `g3_time_convert`) a function of the number of characters in year and step. This means that for `g3s_time`, the `times` argument is replaced with a `time_data` dataframe. This is because  the multiplier is a function of year and step, rather than a lookup function.

lentinj commented on R/stock_time.R:
```diff
@@ -1,51 +1,61 @@
-# - times: Vector of g3s_time_convert(year, step) for each year/step data applies to
-g3s_time <- function(inner_stock, times, year = NULL, step = NULL) {
+# - time_data: dataframe that contains a year and possibly a step column
+g3s_time <- function(inner_stock, time_data, year = NULL, step = NULL) {
```
I see why you've done it, but this interface-change breaks the symmetry between the other g3s_* functions, which all expect a list of unique values one way or another. This is leaking the work of likelihood_data into g3s_time, which I don't think should happen.

willbutler42 replied:
Yep, this is back to normal with the second commit.

lentinj reviewed (changes requested):
I think the answer here is to simplify. This business with ``mult`` was mostly to keep me sane whilst reading the raw numbers---``202002`` is hard work on your eyes. However, nowadays all the axes have proper labels, so you shouldn't be looking at the raw integers that much.

I'd suggest:
* ``mult`` is always ``10``, regardless of the existence of the step column or number of digits in the year.
* Any ``step > 99`` is an error (should be enough for anyone) 
* In ``g3s_time_labels``(), don't print a step if ``step == 0``.

Hopefully this should fix the problem by mostly throwing code away.

Another option is to investigate the tuples used by ``g3_param_table``. Having a key of ``(year, step)`` is the correct way to do this, but there's not a generic framework for using them, unlike the intlookup. https://github.com/gadget-framework/gadget3/issues/22 is suggesting doing this.

It'd be a lot bigger job than you're necessarily hoping for, and especially in the R case not particularly fast---the closest equivalent to a tuple in R I could find is ``paste(year, step, collapse = '.')``. I think it can wait for another day.

willbutler42 commented:
> I think the answer here is to simplify. This business with `mult` was mostly to keep me sane whilst reading the raw numbers---`202002` is hard work on your eyes. However, nowadays all the axes have proper labels, so you shouldn't be looking at the raw integers that much.
> 
> I'd suggest:
> 
>     * `mult` is always `10`, regardless of the existence of the step column or number of digits in the year.
> 
>     * Any `step > 99` is an error (should be enough for anyone)
> 
>     * In `g3s_time_labels`(), don't print a step if `step == 0`.
> 
> 
> Hopefully this should fix the problem by mostly throwing code away.
> 
> Another option is to investigate the tuples used by `g3_param_table`. Having a key of `(year, step)` is the correct way to do this, but there's not a generic framework for using them, unlike the intlookup. #22 is suggesting doing this.
> 
> It'd be a lot bigger job than you're necessarily hoping for, and especially in the R case not particularly fast---the closest equivalent to a tuple in R I could find is `paste(year, step, collapse = '.')`. I think it can wait for another day.

Thanks, @lentinj . I will follow the fixed mult suggestion and test it out. Agree, the tuples can wait for another day.

lentinj commented:
Sounds good. And sorry it took so long to get back to you!

willbutler42 commented:
This should be fixed now. Following your suggestion, mult is now fixed to 100 and if step is > 99 an error is issued. I removed g3s_time_multiplier though maybe that could be turned into a parameter?

lentinj commented on R/stock_time.R:
```diff
@@ -1,111 +1,94 @@
-        as.character(times)
-    }
+  ## Steps are included
+  if (!all((times / 100) %% 1 == 0)) {
```
Minor nitpick: ``!all()`` in the general case is dramatically slower than ``any()``, as the former will have to iterate over everything regardless, but ``any`` can stop as soon as it finds something. How true this is with R's vectorisation I'm not sure, but it's a good habit.

But more importantly I'm not convinced by the inner part. You're roughly asking "does ``times / 100`` have a fractional part", which as well as being slower by invoking floats, there's a chance the equality won't do what you want. One should also heed the warning in the ``%%`` man page.

Anyway, ``any(times %% 100 > 0)`` is what you really want here IMO. The modulo throws away everything > 100, if anything is > 0 then there's steps.

willbutler42 replied:
Thanks, updated : )

lentinj commented on R/likelihood_data.R:
```diff
@@ -154,6 +154,9 @@ g3l_likelihood_data <- function (nll_name, data, missing_val = 0, area_group = N
     }
+    if ('step' %in% names(data) & any(data$step > 99)) {
+        stop("The number of steps per year cannot exceed 99")
+    }
```
This might make more sense at the top of ``g3s_time_convert``(), since then it's right next to the code that needs this. But I'm just making comments to avoid work now :)

lentinj commented:
> I removed g3s_time_multiplier though maybe that could be turned into a parameter?

I'm not convinced it's worth the faff. You could have a ``stock_time_mult <- 100`` constant at the top of the file which might kee things a bit more readable, and then it's only one place to update if it really needs changing.

willbutler42 commented:
> > I removed g3s_time_multiplier though maybe that could be turned into a parameter?
> 
> I'm not convinced it's worth the faff. You could have a `stock_time_mult <- 100` constant at the top of the file which might kee things a bit more readable, and then it's only one place to update if it really needs changing.

Just left it as is.
